src/ui.cpp

```diff
@@ -639,17 +641,165 @@ void uilist::calc_data()
         } while( !assigned && hotkey != input_event() );
     }
 
-    vmax = entries.size();
+    // To figure out the layout we do two passes over the contents - first to calculate width,
+    // and only then to figure out the height, given the width restrictions.
+    // This is neccessary because we need word wrapping for useful UI,
+    // but word wrapping means that text height is affected by text width.
+
+    ImVec2 available_bounds = ImGui::GetMainViewport()->Size;
+    available_bounds.x *= 0.9; // for aesthetic reasons
+    if( desired_bounds.has_value() ) {
+        // TODO: this should never happen actually, and is a bad API
+        cataimgui::bounds b = desired_bounds.value();
+        if( b.w > 0 ) {
+            available_bounds.x = std::min( available_bounds.x, b.w );
+        }
+        if( b.h > 0 ) {
+            available_bounds.y = std::min( available_bounds.y, b.w );
+        }
+    }
 
-    ImVec2 title_size = ImVec2();
-    bool has_titlebar = !title.empty();
-    if( has_titlebar ) {
+    //
+    // TODO: calc_size does not account for items spacing
+    //
+
+    ImVec2 uilist_size = ImVec2(); // The outer size of the box (with padding)
+
+    // These values hold the inner sizes of elements, i.e. without padding.
+    ImVec2 title_size = ImVec2(); // titlebar, the one with the "x to close"
+    ImVec2 header_size = ImVec2(); // Header, below titlebar
+    ImVec2 menu_size = ImVec2(); // the selectable list of options
+    ImVec2 footer_size = ImVec2(); // static footer
+    ImVec2 desc_size = ImVec2();  // description of the current entry
+    // the three "columns" in the table.
+    ImVec2 hotkey_size = ImGui::CalcTextSize( "M" );
+    ImVec2 label_size = ImVec2(); // first column, label
+    ImVec2 label2_size = ImVec2(); // second column, optional (?)    
+
+    // TODO: this is bad
+    extra_space_left = 0.0;
+    extra_space_right = 0.0;
+    if( callback != nullptr ) {
+        extra_space_left = callback->desired_extra_space_left() + s.FramePadding.x;
+        extra_space_right = callback->desired_extra_space_right() + s.FramePadding.x;
+    }
+
+    if( !title.empty() ) {
+        title_size.x = calc_size( title ).x;
+    }
+    if( !text.empty() ) {
+        header_size.x = calc_size( text ).x;
+    }
+    if( !footer_text.empty() ) {
+        footer_size.x = calc_size( footer_text ).x;
+    }
+    for( const uilist_entry &ent : entries ) {
+        ImVec2 entry_text_size = calc_size( ent.txt );
+        ImVec2 entry_context_size = calc_size( ent.ctxt );
+        ImVec2 entry_desc_size = calc_size( ent.desc );
+
+        label_size.x = std::max( label_size.x, entry_text_size.x );
+        label2_size.x = std::max( label2_size.x, entry_context_size.x );
+        desc_size.x = std::max( desc_size.x, entry_desc_size.x );
+    }
+    if( desc_size.x == 0 ) {
+        desc_enabled = false; // TODO: really?
```

Comment:
translation: maybe there's a better way to disable it than checking that the calculated width is 0 (it might also be fine, idk)

Reply:
Fair enough.